### PR TITLE
add non-null to store non-default values when patching security config

### PR DESCRIPTION
**Pull request by nsri19**
##  opensearch-security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Bug fix
 
 
 2. __Github Issue # or road-map entry, if available:__
#1435 

 3. __Description of changes:__

- Added `JsonInclude.Include.NON_NULL` to `ConfigV6` and `ConfigV7` properties to store all non-null & non-default values
- Added unit test
- Modified `ConfigV6Test` and `ConfigV7Test` as bug causing `multitenancy_enabled` to revert to default values is fixed

 4. __Why these changes are required?__ 
Currently, security config settings that default to true are flipped from false to default (true) on patch requests. This fix removes this bug.
The similar fixes have been applied to `audit/config`: https://github.com/opensearch-project/security/pull/652

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/resources/restapi/securityconfig_nondefault.json [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        "description" : "Authorize via LDAP or Active Directory"
+      }
+    },
+    "auth_failure_listeners" : { },
+    "do_not_fail_on_forbidden" : false,
+    "multi_rolespan_enabled" : true,
+    "hosts_resolver_mode" : "ip-only",
+    "do_not_fail_on_forbidden_empty" : false
+  }
+}
```
Comment (cliu123) [inline src/test/resources/restapi/securityconfig_nondefault.json:173]: Please add an empty line at EOF.

File: src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java [hunk 2]
```diff
@@ -40,6 +40,7 @@
 import com.fasterxml.jackson.annotation.JsonAnyGetter;
 import com.fasterxml.jackson.annotation.JsonAnySetter;
 import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
```
Comment (cliu123) [inline src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java:43]: How is the changes applied to `ConfigV6` tested?

File: src/test/resources/restapi/securityconfig_nondefault.json (hunk 1 above)
Comment (nsri19) [reply src/test/resources/restapi/securityconfig_nondefault.json]: done, thanks!

File: src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java (hunk 2 above)
Comment (nsri19) [reply src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java:43]: Not able to test changes applied to `ConfigV6`. Keeping the changes to maintain consistency with `ConfigV7`.

**Comment by vrozov** [comment]
Is security config the only impacted object? Can the same apply to other patchable objects? If other objects are also impacted, should they be also changed? Or it may be better to change `ObjectMapper` behavior globally? Another option to consider is to apply annotation only to properties impacted (reverted) by patch.

File: src/test/resources/restapi/securityconfig_nondefault.json (hunk 1 above)
Comment (nsri19) [reply src/test/resources/restapi/securityconfig_nondefault.json]: done, thank you!

File: src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java (hunk 2 above)
Comment (hardik-k-shah) [reply src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java:43]: please create 6x domain to test those changes.

File: src/test/java/org/opensearch/security/dlic/rest/api/SecurityConfigApiTest.java [hunk 3]
```diff
@@ -108,4 +109,37 @@ public void testSecurityConfigApiWrite() throws Exception {
         Assert.assertEquals(HttpStatus.SC_METHOD_NOT_ALLOWED, response.getStatusCode());
 
     }
+
+    @Test
+    public void testSecurityConfigPatch() throws Exception {
```
Comment (hardik-k-shah) [inline src/test/java/org/opensearch/security/dlic/rest/api/SecurityConfigApiTest.java:114]: testSecurityConfigForHTTPPatch()

**Review by hardik-k-shah — COMMENTED** [review]
We should use `HTTP PATCH method request` instead `patching` to avoid confusion with software patching procedure.

File: src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java (hunk 2 above)
Comment (hardik-k-shah) [reply src/main/java/org/opensearch/security/securityconf/impl/v6/ConfigV6.java:43]: As discussed, this is mainly for 6.x to 7.x migration and will not impact 6.x directly. But lets test migration case as discussed offline.

**Review by hardik-k-shah — APPROVED** [review]

**Review by davidlago — APPROVED** [review]

**Pull request merged**